# Part 1 - Data Cleaning


## RECAP: The Business Problem
---
### OVERVIEW
Property owners who wish to put their property up for sale tend to want to obtain a preliminary valuation for consideration. 

As a Data Scientist for a Property-Tech Company in the USA, we are tasked to build the algorithm behind this web app and recommend the optimal number of features to provide a balance between accuracy and user experience (because nobody likes to fill in lengthy forms!). Having this app is also important to attract users to use our online property marketplace and allow us to earn revenue from partnerships & ads.

To ensure that User Experience is optimized, the user is not going to input more than 10 fields in our web app as it would increase the cost of engagement of the user and may result in lower utilisation of our web app.

### PROBLEM STATEMENT

Using the regression models learnt (Linear Regression, Lasso, Ridge, Elastic Net), decide on a regression model that will be deployed in our web app that uses not more than 10 features (which corresponds to 10 fields) within a reasonable RMSE to provide a prediction on a property's selling price.

We will also be providing submissions to Kaggle to score our regression models against unseen data.

## Housekeeping


We will import the libraries we need and get acquantined with our dataset.


The Ames Housing Dataset has **2051 rows** and **81 features** and there are multiple columns which have missing (null) values.


The objective of this notebook is to identify and impute missing values with the appropriate values. The data will first be split up into the different data types and cleaned respectively.
- Ordinal Data
- Discrete Data
- Continuous Data

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# load csv 

pd.options.display.max_columns = 100

df = pd.read_csv(".../datasets/train.csv")
display(df.head())
display(df.shape)
display(df.info())

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0  109  533352170           60        RL           NaN     13517   Pave   NaN   
1  544  531379050           60        RL          43.0     11492   Pave   NaN   
2  153  535304180           20        RL          68.0      7922   Pave   NaN   
3  318  916386060           60        RL          73.0      9802   Pave   NaN   
4  255  906425045           50        RL          82.0     14235   Pave   NaN   

  Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0       IR1          Lvl    AllPub    CulDSac        Gtl       Sawyer   
1       IR1          Lvl    AllPub    CulDSac        Gtl      SawyerW   
2       Reg          Lvl    AllPub     Inside        Gtl        NAmes   
3       Reg          Lvl    AllPub     Inside        Gtl       Timber   
4       IR1          Lvl    AllPub     Inside        Gtl      SawyerW   

  Condition 1 Condition 2 Bldg Type House Style  Overall Qual  Overall Cond  \
0        RRAe        Norm      1Fam      2Story             6             8   
1        Norm        Norm      1Fam      2Story             7             5   
2        Norm        Norm      1Fam      1Story             5             7   
3        Norm        Norm      1Fam      2Story             5             5   
4        Norm        Norm      1Fam      1.5Fin             6             8   

   Year Built  Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd  \
0        1976            2005      Gable   CompShg      HdBoard      Plywood   
1        1996            1997      Gable   CompShg      VinylSd      VinylSd   
2        1953            2007      Gable   CompShg      VinylSd      VinylSd   
3        2006            2007      Gable   CompShg      VinylSd      VinylSd   
4        1900            1993      Gable   CompShg      Wd Sdng      Plywood   

  Mas Vnr Type  Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual  \
0      BrkFace         289.0         Gd         TA     CBlock        TA   
1      BrkFace         132.0         Gd         TA      PConc        Gd   
2         None           0.0         TA         Gd     CBlock        TA   
3         None           0.0         TA         TA      PConc        Gd   
4         None           0.0         TA         TA      PConc        Fa   

  Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  \
0        TA            No            GLQ         533.0            Unf   
1        TA            No            GLQ         637.0            Unf   
2        TA            No            GLQ         731.0            Unf   
3        TA            No            Unf           0.0            Unf   
4        Gd            No            Unf           0.0            Unf   

   BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air  \
0           0.0        192.0          725.0    GasA         Ex           Y   
1           0.0        276.0          913.0    GasA         Ex           Y   
2           0.0        326.0         1057.0    GasA         TA           Y   
3           0.0        384.0          384.0    GasA         Gd           Y   
4           0.0        676.0          676.0    GasA         TA           Y   

  Electrical  1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  \
0      SBrkr         725         754                0         1479   
1      SBrkr         913        1209                0         2122   
2      SBrkr        1057           0                0         1057   
3      SBrkr         744         700                0         1444   
4      SBrkr         831         614                0         1445   

   Bsmt Full Bath  Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  \
0             0.0             0.0          2          1              3   
1             1.0             0.0          2          1              4   
2             1.0             0.0          1          0              3   
3             0.0             0.0          2  

(2051, 81)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2051 entries, 0 to 2050
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               2051 non-null   int64  
 1   PID              2051 non-null   int64  
 2   MS SubClass      2051 non-null   int64  
 3   MS Zoning        2051 non-null   object 
 4   Lot Frontage     1721 non-null   float64
 5   Lot Area         2051 non-null   int64  
 6   Street           2051 non-null   object 
 7   Alley            140 non-null    object 
 8   Lot Shape        2051 non-null   object 
 9   Land Contour     2051 non-null   object 
 10  Utilities        2051 non-null   object 
 11  Lot Config       2051 non-null   object 
 12  Land Slope       2051 non-null   object 
 13  Neighborhood     2051 non-null   object 
 14  Condition 1      2051 non-null   object 
 15  Condition 2      2051 non-null   object 
 16  Bldg Type        2051 non-null   object 
 17  House Style   

None

## Splitting the Data into the Various Data Types


The columns were examined manually and categorized into the 3 categories. We will also create a list to track the columns dropped (if required).

In [2]:
# Manual sorting of variables

ordinal = ["Lot Shape", "Land Slope", "Overall Qual", "Overall Cond", "Exter Qual", "Exter Cond", 
    "Bsmt Qual", "Bsmt Cond", "Bsmt Exposure", "BsmtFin Type 1", "BsmtFin Type 2", "Heating QC", 
    "Central Air", "Kitchen Qual", "Functional", "Fireplace Qu", "Garage Finish", "Garage Qual", 
    "Garage Cond", "Paved Drive", "Pool QC", "Fence"]

discrete = ["Year Built", "Year Remod/Add", "Bsmt Full Bath", "Bsmt Half Bath", "Full Bath", "Half Bath",
    "Bedroom AbvGr", "Kitchen AbvGr", "TotRms AbvGrd", "Fireplaces", "Garage Yr Blt", "Garage Cars", "Mo Sold", "Yr Sold"]

continuous = ["Lot Frontage", "Lot Area", "Mas Vnr Area", "BsmtFin SF 1", "BsmtFin SF 2", 
    "Bsmt Unf SF", "Total Bsmt SF", "1st Flr SF", "2nd Flr SF", "Low Qual Fin SF", "Gr Liv Area", 
    "Garage Area", "Wood Deck SF", "Open Porch SF", "Enclosed Porch", "3Ssn Porch", "Screen Porch", "Pool Area",
    "Misc Val"]

nominal = ["MS SubClass", "MS Zoning", "Street", "Alley", "Land Contour", "Utilities", "Lot Config", "Neighborhood", "Condition 1", "Condition 2", 
    "Bldg Type", "House Style", "Roof Style", "Roof Matl", "Exterior 1st", "Exterior 2nd", "Mas Vnr Type", 
    "Foundation", "Heating", "Electrical", "Garage Type", "Misc Feature", "Sale Type"]

columns_dropped = [] # To keep a record of columns we dropped so we can do drop the same column in the test set

## Cleaning the Ordinal Features

From the Data Dictionary, we know that the missing values was because the property did not possess the particular feature (i.e. fence, pool, basement, garage, fireplace). 


The null features (with the exception of `Bsmt Exposure` and `BsmtFin Type 2`) can be imputed with "none".

In [3]:
df[ordinal].isnull().sum()

Lot Shape            0
Land Slope           0
Overall Qual         0
Overall Cond         0
Exter Qual           0
Exter Cond           0
Bsmt Qual           55
Bsmt Cond           55
Bsmt Exposure       58
BsmtFin Type 1      55
BsmtFin Type 2      56
Heating QC           0
Central Air          0
Kitchen Qual         0
Functional           0
Fireplace Qu      1000
Garage Finish      114
Garage Qual        114
Garage Cond        114
Paved Drive          0
Pool QC           2042
Fence             1651
dtype: int64

In [4]:
df["Fireplace Qu"].fillna("none", inplace=True)
df["Garage Finish"].fillna("none", inplace=True)
df["Garage Qual"].fillna("none", inplace=True)
df["Garage Cond"].fillna("none", inplace=True)
df["Pool QC"].fillna("none", inplace=True)
df["Fence"].fillna("none", inplace=True)

In [5]:
df[ordinal].isnull().sum()

Lot Shape          0
Land Slope         0
Overall Qual       0
Overall Cond       0
Exter Qual         0
Exter Cond         0
Bsmt Qual         55
Bsmt Cond         55
Bsmt Exposure     58
BsmtFin Type 1    55
BsmtFin Type 2    56
Heating QC         0
Central Air        0
Kitchen Qual       0
Functional         0
Fireplace Qu       0
Garage Finish      0
Garage Qual        0
Garage Cond        0
Paved Drive        0
Pool QC            0
Fence              0
dtype: int64

### Basement Features


There were likely 55 properties with no basements yet there were 58 and 56 null values for `Bsmt Exposure` and `BsmtFin Type 2`, the erroneous rows will be examined in context with the other basement features to ensure that it was not a mistake.


We will first define a function which takes 2 DataFrame column names and return the indexes of the excess null rows so that we can drop them. It also compares the indexes to ensure that the indexes in the columns are identical.

In [6]:
def compare_elements(first_col, second_col): 
    
    '''compares whether the cols of null value indexes are identical and returns indexes of the excess rows if any'''
    '''assumes that items within each index list is unique '''
    
    first_index = df[df[first_col].isnull()].index.sort_values()
    second_index = df[df[second_col].isnull()].index.sort_values()
    excess =[]
    
    if len(first_index) == len(second_index):
        if ((first_index == second_index).all()):
            print("Indexes are identical\n")
    else:
        for item in second_index: 
            if item not in first_index:
                excess.append(item)
        print("There are", len(excess), "excess item/s.\n")
        if len(excess) > 0:
              for item in excess:
                print("The excess item/s are:", item, "\n")

In [7]:
bsmt_features = ["Bsmt Qual", "Bsmt Cond", "Bsmt Exposure", "BsmtFin Type 1", "BsmtFin SF 1", "BsmtFin Type 2", "BsmtFin SF 2", 
                 "Bsmt Unf SF", "Total Bsmt SF", "Bsmt Half Bath", "Bsmt Full Bath"]

# For Bsmt Exposure
compare_elements("Bsmt Qual", "Bsmt Exposure")

display(df.iloc[1456][bsmt_features])
display(df.iloc[1547][bsmt_features])
display(df.iloc[1997][bsmt_features])

There are 3 excess item/s.

The excess item/s are: 1456 

The excess item/s are: 1547 

The excess item/s are: 1997 



Bsmt Qual            Gd
Bsmt Cond            TA
Bsmt Exposure       NaN
BsmtFin Type 1      Unf
BsmtFin SF 1        0.0
BsmtFin Type 2      Unf
BsmtFin SF 2        0.0
Bsmt Unf SF       725.0
Total Bsmt SF     725.0
Bsmt Half Bath      0.0
Bsmt Full Bath      0.0
Name: 1456, dtype: object

Bsmt Qual             Gd
Bsmt Cond             TA
Bsmt Exposure        NaN
BsmtFin Type 1       Unf
BsmtFin SF 1         0.0
BsmtFin Type 2       Unf
BsmtFin SF 2         0.0
Bsmt Unf SF       1595.0
Total Bsmt SF     1595.0
Bsmt Half Bath       0.0
Bsmt Full Bath       0.0
Name: 1547, dtype: object

Bsmt Qual            Gd
Bsmt Cond            TA
Bsmt Exposure       NaN
BsmtFin Type 1      Unf
BsmtFin SF 1        0.0
BsmtFin Type 2      Unf
BsmtFin SF 2        0.0
Bsmt Unf SF       936.0
Total Bsmt SF     936.0
Bsmt Half Bath      0.0
Bsmt Full Bath      0.0
Name: 1997, dtype: object

These are erroneous entries as a NaN value for "Bsmt Exposure" means there is **NO** basement. It is definitely not the case as the other basement features are filled. A reasonable value would be "No" - indicating no basement exposure and will be imputed for the above 3 rows. This is done in the cell below.

In [8]:
df.iloc[1456,33]  = "No"
df.iloc[1547,33]  = "No"
df.iloc[1997,33]  = "No"

In [9]:
# For BsmtFin Type 2

compare_elements("Bsmt Qual", "BsmtFin Type 2")
display(df.iloc[1147][bsmt_features])

There are 1 excess item/s.

The excess item/s are: 1147 



Bsmt Qual             Gd
Bsmt Cond             TA
Bsmt Exposure         No
BsmtFin Type 1       GLQ
BsmtFin SF 1      1124.0
BsmtFin Type 2       NaN
BsmtFin SF 2       479.0
Bsmt Unf SF       1603.0
Total Bsmt SF     3206.0
Bsmt Half Bath       0.0
Bsmt Full Bath       1.0
Name: 1147, dtype: object

Since the other basement features seem to suggest that it is an unfinished basement, having NaN as the BsmtFin Type 2 is not correct, it should be "UnF" and will be imputed in the cell below.


After addressing the errorneous entries, we will also impute "none" for the properties that do not have a basement.

In [10]:
df.iloc[1147, 36] = "Unf"

df["Bsmt Qual"].fillna("none", inplace=True)
df["Bsmt Cond"].fillna("none", inplace=True)
df["Bsmt Exposure"].fillna("none", inplace=True)
df["BsmtFin Type 1"].fillna("none", inplace=True)
df["BsmtFin Type 2"].fillna("none", inplace=True)

In [11]:
display(df[ordinal].isnull().sum())

Lot Shape         0
Land Slope        0
Overall Qual      0
Overall Cond      0
Exter Qual        0
Exter Cond        0
Bsmt Qual         0
Bsmt Cond         0
Bsmt Exposure     0
BsmtFin Type 1    0
BsmtFin Type 2    0
Heating QC        0
Central Air       0
Kitchen Qual      0
Functional        0
Fireplace Qu      0
Garage Finish     0
Garage Qual       0
Garage Cond       0
Paved Drive       0
Pool QC           0
Fence             0
dtype: int64

## Cleaning Discrete Features


We will examine the null columns in the context of the related features to decide the value to be imputed.

In [12]:
df[discrete].isnull().sum()

Year Built          0
Year Remod/Add      0
Bsmt Full Bath      2
Bsmt Half Bath      2
Full Bath           0
Half Bath           0
Bedroom AbvGr       0
Kitchen AbvGr       0
TotRms AbvGrd       0
Fireplaces          0
Garage Yr Blt     114
Garage Cars         1
Mo Sold             0
Yr Sold             0
dtype: int64

### Basement Half/Full Baths


The null values are for properties with no basements, we will impute the null values with "none"

In [13]:
display(df[df["Bsmt Full Bath"].isnull()][bsmt_features])
display(df[df["Bsmt Half Bath"].isnull()][bsmt_features])

df["Bsmt Full Bath"].fillna(0, inplace=True)
df["Bsmt Half Bath"].fillna(0, inplace=True)

display(df[df["Bsmt Full Bath"].isnull()][bsmt_features]) # Verifying values have been imputed
display(df[df["Bsmt Half Bath"].isnull()][bsmt_features]) 

Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1  \
616       none      none          none           none           0.0   
1327      none      none          none           none           NaN   

     BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  Bsmt Half Bath  \
616            none           0.0          0.0            0.0             NaN   
1327           none           NaN          NaN            NaN             NaN   

      Bsmt Full Bath  
616              NaN  
1327             NaN

Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1  \
616       none      none          none           none           0.0   
1327      none      none          none           none           NaN   

     BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  Bsmt Half Bath  \
616            none           0.0          0.0            0.0             NaN   
1327           none           NaN          NaN            NaN             NaN   

      Bsmt Full Bath  
616              NaN  
1327             NaN

Empty DataFrame
Columns: [Bsmt Qual, Bsmt Cond, Bsmt Exposure, BsmtFin Type 1, BsmtFin SF 1, BsmtFin Type 2, BsmtFin SF 2, Bsmt Unf SF, Total Bsmt SF, Bsmt Half Bath, Bsmt Full Bath]
Index: []

Empty DataFrame
Columns: [Bsmt Qual, Bsmt Cond, Bsmt Exposure, BsmtFin Type 1, BsmtFin SF 1, BsmtFin Type 2, BsmtFin SF 2, Bsmt Unf SF, Total Bsmt SF, Bsmt Half Bath, Bsmt Full Bath]
Index: []

### Garage Yr Built & Garage Cars


For `Garage Yr Blt`, the null values are for the properties with no garages. "none" will be imputed.


For `Garage Cars`, it is likely an erroneous entry as most garage features are null values. "none" will be imputed.

In [14]:
garage_features = ["Garage Type", "Garage Yr Blt", "Garage Finish", "Garage Cars", "Garage Area", "Garage Qual", "Garage Cond"]
display(df[df["Garage Yr Blt"].isnull()][garage_features])
display(df[df["Garage Cars"].isnull()][garage_features])

df["Garage Yr Blt"].fillna(0, inplace=True)
df["Garage Cars"].fillna(0, inplace=True)

display(df[df["Garage Yr Blt"].isnull()][garage_features]) # verifying that values have been imputed
display(df[df["Garage Cars"].isnull()][garage_features])

Garage Type  Garage Yr Blt Garage Finish  Garage Cars  Garage Area  \
28           NaN            NaN          none          0.0          0.0   
53           NaN            NaN          none          0.0          0.0   
65           NaN            NaN          none          0.0          0.0   
79           NaN            NaN          none          0.0          0.0   
101          NaN            NaN          none          0.0          0.0   
...          ...            ...           ...          ...          ...   
1991         NaN            NaN          none          0.0          0.0   
2010         NaN            NaN          none          0.0          0.0   
2027         NaN            NaN          none          0.0          0.0   
2039         NaN            NaN          none          0.0          0.0   
2042         NaN            NaN          none          0.0          0.0   

     Garage Qual Garage Cond  
28          none        none  
53          none        none  
65          none        none  
79          none        none  
101         none        none  
...          ...         ...  
1991        none        none  
2010        none        none  
2027        none        none  
2039        none        none  
2042        none        none  

[114 rows x 7 columns]

Garage Type  Garage Yr Blt Garage Finish  Garage Cars  Garage Area  \
1712      Detchd            NaN          none          NaN          NaN   

     Garage Qual Garage Cond  
1712        none        none

Empty DataFrame
Columns: [Garage Type, Garage Yr Blt, Garage Finish, Garage Cars, Garage Area, Garage Qual, Garage Cond]
Index: []

Empty DataFrame
Columns: [Garage Type, Garage Yr Blt, Garage Finish, Garage Cars, Garage Area, Garage Qual, Garage Cond]
Index: []

In [15]:
df[discrete].isnull().sum()

Year Built        0
Year Remod/Add    0
Bsmt Full Bath    0
Bsmt Half Bath    0
Full Bath         0
Half Bath         0
Bedroom AbvGr     0
Kitchen AbvGr     0
TotRms AbvGrd     0
Fireplaces        0
Garage Yr Blt     0
Garage Cars       0
Mo Sold           0
Yr Sold           0
dtype: int64

## Cleaning the Continuous Features


Using domain knowledge, for `Lot Frontage` values, it is highly dependent on the site of the property (i.e. the other features) and its effect on SalePrice will be assumed to be captured by the other features. We will drop it for simplicity.

In [16]:
display(df[continuous].isnull().sum())
df.drop("Lot Frontage", axis=1, inplace=True)

columns_dropped = columns_dropped + ["Lot Frontage"]

Lot Frontage       330
Lot Area             0
Mas Vnr Area        22
BsmtFin SF 1         1
BsmtFin SF 2         1
Bsmt Unf SF          1
Total Bsmt SF        1
1st Flr SF           0
2nd Flr SF           0
Low Qual Fin SF      0
Gr Liv Area          0
Garage Area          1
Wood Deck SF         0
Open Porch SF        0
Enclosed Porch       0
3Ssn Porch           0
Screen Porch         0
Pool Area            0
Misc Val             0
dtype: int64

### Mas Vnr Area


Correlating to the related feature (`Mas Vnr Type`) and the data dictionary, the null values in `Mas Vnr Area` are due to properties who did not have masonry veneer. We will impute it with 0.

In [17]:
display(df[df["Mas Vnr Area"].isnull()]["Mas Vnr Type"])

df["Mas Vnr Area"].fillna(0, inplace=True)

display(df["Mas Vnr Area"].isnull().sum()) # verify that value has been imputed

22      NaN
41      NaN
86      NaN
212     NaN
276     NaN
338     NaN
431     NaN
451     NaN
591     NaN
844     NaN
913     NaN
939     NaN
1025    NaN
1244    NaN
1306    NaN
1430    NaN
1434    NaN
1606    NaN
1699    NaN
1815    NaN
1820    NaN
1941    NaN
Name: Mas Vnr Type, dtype: object

0

### Basement Features


Examination of the missing Basement continuous features yield that the null values come from the same row. Impute the value of 0 as this property does not have a basement.

In [18]:
display(df[df["BsmtFin SF 1"].isnull()][bsmt_features]) 
display(df[df["BsmtFin SF 2"].isnull()][bsmt_features])
display(df[df["Bsmt Unf SF"].isnull()][bsmt_features])
display(df[df["Total Bsmt SF"].isnull()][bsmt_features]) # Index 1327

df["BsmtFin SF 1"].fillna(0, inplace=True)
df["BsmtFin SF 2"].fillna(0, inplace=True)
df["Bsmt Unf SF"].fillna(0, inplace=True)
df["Total Bsmt SF"].fillna(0, inplace=True)

df.iloc[1327][bsmt_features] # verifying that values are imputed

Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1  \
1327      none      none          none           none           NaN   

     BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  Bsmt Half Bath  \
1327           none           NaN          NaN            NaN             0.0   

      Bsmt Full Bath  
1327             0.0

Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1  \
1327      none      none          none           none           NaN   

     BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  Bsmt Half Bath  \
1327           none           NaN          NaN            NaN             0.0   

      Bsmt Full Bath  
1327             0.0

Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1  \
1327      none      none          none           none           NaN   

     BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  Bsmt Half Bath  \
1327           none           NaN          NaN            NaN             0.0   

      Bsmt Full Bath  
1327             0.0

Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1  \
1327      none      none          none           none           NaN   

     BsmtFin Type 2  BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF  Bsmt Half Bath  \
1327           none           NaN          NaN            NaN             0.0   

      Bsmt Full Bath  
1327             0.0

Bsmt Qual         none
Bsmt Cond         none
Bsmt Exposure     none
BsmtFin Type 1    none
BsmtFin SF 1       0.0
BsmtFin Type 2    none
BsmtFin SF 2       0.0
Bsmt Unf SF        0.0
Total Bsmt SF      0.0
Bsmt Half Bath     0.0
Bsmt Full Bath     0.0
Name: 1327, dtype: object

### Garage Area

The null value comes from the erroneous entry we encountered when we imputed the discrete features earlier. Imputed value will be 0.

In [19]:
display(df[df["Garage Area"].isnull()][garage_features])

df["Garage Area"].fillna(0, inplace=True)

display(df[df["Garage Area"].isnull()][garage_features]) # verify that value is imputed

Garage Type  Garage Yr Blt Garage Finish  Garage Cars  Garage Area  \
1712      Detchd            0.0          none          0.0          NaN   

     Garage Qual Garage Cond  
1712        none        none

Empty DataFrame
Columns: [Garage Type, Garage Yr Blt, Garage Finish, Garage Cars, Garage Area, Garage Qual, Garage Cond]
Index: []

In [20]:
df[["Lot Area", "Mas Vnr Area", "BsmtFin SF 1", "BsmtFin SF 2", 
"Bsmt Unf SF", "Total Bsmt SF", "1st Flr SF", "2nd Flr SF", "Low Qual Fin SF", "Gr Liv Area", 
"Garage Area", "Wood Deck SF", "Open Porch SF", "Enclosed Porch", "3Ssn Porch", "Screen Porch", "Pool Area",
"Misc Val"]].isnull().sum()

Lot Area           0
Mas Vnr Area       0
BsmtFin SF 1       0
BsmtFin SF 2       0
Bsmt Unf SF        0
Total Bsmt SF      0
1st Flr SF         0
2nd Flr SF         0
Low Qual Fin SF    0
Gr Liv Area        0
Garage Area        0
Wood Deck SF       0
Open Porch SF      0
Enclosed Porch     0
3Ssn Porch         0
Screen Porch       0
Pool Area          0
Misc Val           0
dtype: int64

## Cleaning Nominal Features

For `Alley` and `Misc Feature`, the properties do not have these features, imputed value will be "none".


Referencing the Data Dictionary, for `Mas Vnr Type`, the properties did not have Masonry Veneers, imputed value will be "none"


Reference to the cleaning of ordinal features, we know that 114 properties do not have a garage (instead of 113 here). This is due to index 1712 which we encountered earlier.

In [21]:
display(df[nominal].isnull().sum())

display(df.iloc[1712][garage_features])

MS SubClass        0
MS Zoning          0
Street             0
Alley           1911
Land Contour       0
Utilities          0
Lot Config         0
Neighborhood       0
Condition 1        0
Condition 2        0
Bldg Type          0
House Style        0
Roof Style         0
Roof Matl          0
Exterior 1st       0
Exterior 2nd       0
Mas Vnr Type      22
Foundation         0
Heating            0
Electrical         0
Garage Type      113
Misc Feature    1986
Sale Type          0
dtype: int64

Garage Type      Detchd
Garage Yr Blt       0.0
Garage Finish      none
Garage Cars         0.0
Garage Area         0.0
Garage Qual        none
Garage Cond        none
Name: 1712, dtype: object

In [22]:
df["Mas Vnr Type"].value_counts()

None       1218
BrkFace     630
Stone       168
BrkCmn       13
Name: Mas Vnr Type, dtype: int64

In [23]:
df["Alley"].fillna("none", inplace=True)
df["Misc Feature"].fillna("none", inplace=True)
df["Garage Type"].fillna("none", inplace=True)
df["Mas Vnr Type"].fillna("None", inplace=True)

display(df[nominal].isnull().sum()) # Verify values are imputed

MS SubClass     0
MS Zoning       0
Street          0
Alley           0
Land Contour    0
Utilities       0
Lot Config      0
Neighborhood    0
Condition 1     0
Condition 2     0
Bldg Type       0
House Style     0
Roof Style      0
Roof Matl       0
Exterior 1st    0
Exterior 2nd    0
Mas Vnr Type    0
Foundation      0
Heating         0
Electrical      0
Garage Type     0
Misc Feature    0
Sale Type       0
dtype: int64

In [24]:
display(df.isnull().sum().sum()) # verify that all null values were dealth with
display(df.shape)

0

(2051, 80)

## Conclusion for Part I


In this notebook, we have addressed and imputed the null values for the respective data types. We have also dropped the `Lot Frontage` feature.


The cleaned dataset is exported and will be used in Part II to conduct Exploratory Data Analysis.


The dataset now has 80 columns instead of 81 columns.

In [25]:
df.to_csv("df_clean.csv", index_label = False) # export the cleaned dataset
display(columns_dropped)
display(df.shape)

['Lot Frontage']

(2051, 80)